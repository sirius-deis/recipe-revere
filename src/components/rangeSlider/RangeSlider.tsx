import { FC, useCallback, useEffect, useRef, useState } from "react";
import styles from "./RangleSlider.module.css";

interface RangeSliderProps {
  min: number;
  max: number;
  onChange: (minVal: number, maxVal: number) => void;
  width?: number;
  withLabel?: boolean;
}

const RangeSlider: FC<RangeSliderProps> = ({
  max,
  min,
  onChange,
  width,
  withLabel = true,
}) => {
  const [minVal, setMinVal] = useState(min);
  const [maxVal, setMaxVal] = useState(max);
  const minValueRef = useRef(min);
  const maxValueRef = useRef(max);
  const range = useRef(null);

  const getValueInPercent = useCallback(
    (value: number) => Math.round(((value - min) / (max - min)) * 100),
    [min, max]
  );

  useEffect(() => {
    const minPercentage = getValueInPercent(minVal);
    const maxPercentage = getValueInPercent(maxValueRef.current);

    if (range.current) {
      (range.current as HTMLDivElement).style.left = `${minPercentage}%`;
      (range.current as HTMLDivElement).style.width = `${
        maxPercentage - minPercentage
      }%`;
    }
  }, [minVal, getValueInPercent]);

  useEffect(() => {
    const minPercentage = getValueInPercent(minValueRef.current);
    const maxPercentage = getValueInPercent(maxVal);

    if (range.current) {
      (range.current as HTMLDivElement).style.width = `${
        maxPercentage - minPercentage
      }%`;
    }
  }, [maxVal, getValueInPercent]);

  useEffect(() => {
    onChange(minVal, maxVal);
  }, [minVal, maxVal, onChange]);

  return (
    <div
      className={styles["range-slider"]}
      style={{ width: width ? `${width}px` : "100%" }}
    >
      <input
        type="range"
        className={`${styles.thumb} ${styles["thumb-left"]}`}
        min={min}
        max={max}
        defaultValue={minVal}
      />
      <input
        type="range"
        className={`${styles.thumb} ${styles["thumb-right"]}`}
        min={min}
        max={max}
        defaultValue={maxVal}
      />
      <div className={styles.slider}>
        <div className={styles["slider-track"]} />
        <div ref={range} className={styles["slider-range"]} />
        {withLabel && (
          <>
            <div
              className={`${styles["slider-value"]} ${styles["slider-value-left"]}`}
            >
              {minVal}
            </div>
            <div
              className={`${styles["slider-value"]} ${styles["slider-value-right"]}`}
            >
              {maxVal}
            </div>
          </>
        )}
      </div>
    </div>
  );
};

export default RangeSlider;
