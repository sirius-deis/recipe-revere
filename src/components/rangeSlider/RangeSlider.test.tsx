import { render, screen, fireEvent } from "@testing-library/react";
import RangeSlider from "./RangeSlider";

describe("RangeSlider component", () => {
  it("should match snapshot", () => {
    const { container } = render(
      <RangeSlider min={0} max={100} onChange={() => {}} />
    );
    expect(container).toMatchSnapshot();
  });
  it("should render component with proper values", () => {
    render(<RangeSlider min={0} max={100} onChange={() => {}} />);
    expect(screen.getByText("0")).toBeInTheDocument();
    expect(screen.getByText("100")).toBeInTheDocument();
  });
  it("should fire event", () => {
    const fn = jest.fn();
    render(<RangeSlider min={0} max={100} onChange={fn} />);
    fireEvent.change(screen.getByTestId("range1"), {
      target: { value: 30 },
    });
    expect(screen.getByText("30")).toBeInTheDocument();
    expect(fn).toHaveBeenCalled();
    expect(fn).toHaveBeenCalledTimes(2);
    expect(fn).toHaveBeenCalledWith(30, 100);

    fireEvent.change(screen.getByTestId("range2"), {
      target: { value: 70 },
    });
    expect(screen.getByText("70")).toBeInTheDocument();
    expect(fn).toHaveBeenCalled();
    expect(fn).toHaveBeenCalledTimes(3);
    expect(fn).toHaveBeenCalledWith(30, 70);
  });
});
