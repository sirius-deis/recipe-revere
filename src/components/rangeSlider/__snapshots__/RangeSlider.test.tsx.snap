// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RangeSlider component should match snapshot 1`] = `
<div>
  <div
    class="range-slider"
    style="width: 100%;"
  >
    <input
      class="thumb thumb-left"
      data-testid="range1"
      max="100"
      min="0"
      type="range"
      value="0"
    />
    <input
      class="thumb thumb-right"
      data-testid="range2"
      max="100"
      min="0"
      type="range"
      value="100"
    />
    <div
      class="slider"
    >
      <div
        class="slider-track"
      />
      <div
        class="slider-range"
        style="left: 0%; width: 100%;"
      />
      <div
        class="slider-value slider-value-left"
      >
        0
      </div>
      <div
        class="slider-value slider-value-right"
      >
        100
      </div>
    </div>
  </div>
</div>
`;
